File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
  constructor(a, b) {
    this.real = a;
    this.imag = b;
  }

  get abs() {
    return Math.sqrt(this.real * this.real + this.imag * this.imag);
  }

  get conj() {
    return new ComplexNumber(this.real, this.imag === 0 ? 0 : -this.imag);
  }

  get exp() {
    return new ComplexNumber(
      Math.E ** this.real * Math.cos(this.imag),
      Math.E ** this.real * Math.sin(this.imag),
    );
  }

  add(cn) {
    return new ComplexNumber(this.real + cn.real, this.imag + cn.imag); 
  }

  sub(cn) {
    return new ComplexNumber(this.real - cn.real, this.imag - cn.imag); 
  }

  mul(cn) {
    return new ComplexNumber(this.real * cn.real - this.imag * cn.imag, this.imag * cn.real + this.real * cn.imag); 
  }

  div(cn) {
    return new ComplexNumber(
      (this.real * cn.real + this.imag * cn.imag) / (cn.real * cn.real + cn.imag * cn.imag),
      (this.imag * cn.real - this.real * cn.imag) / (cn.real * cn.real + cn.imag * cn.imag)
    );
  }
}

File: kotlin/complex-numbers/src/main/kotlin/ComplexNumbers.kt
data class ComplexNumber(val real: Double = 0.0, val imag: Double = 0.0) {
  var abs = Math.sqrt(this.real * this.real + this.imag * this.imag)
  fun conjugate() = ComplexNumber(this.real, 0 - this.imag)
}

operator fun ComplexNumber.plus(other: ComplexNumber) = ComplexNumber(
  this.real + other.real, this.imag + other.imag
)

operator fun ComplexNumber.minus(other: ComplexNumber) = ComplexNumber(
  this.real - other.real, this.imag - other.imag
)

operator fun ComplexNumber.times(other: ComplexNumber) = ComplexNumber(
  this.real * other.real - this.imag * other.imag,
  this.imag * other.real + this.real * other.imag
)

operator fun ComplexNumber.div(other: ComplexNumber) = ComplexNumber(
  (this.real * other.real + this.imag * other.imag) / (other.real * other.real + other.imag * other.imag),
  (this.imag * other.real - this.real * other.imag) / (other.real * other.real + other.imag * other.imag)
)

fun exponential(c: ComplexNumber) = ComplexNumber(
  Math.exp(c.real) * (Math.cos(c.imag) + Math.sin(c.imag)),
  0.0
)